File: SRP Redux/Exercise/MarsRoverKata/Rover.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace MarsRoverKata
{
    public class Rover
    {
        public Point Location { get; set; }
        public Direction Facing { get; set; }

        private static readonly Dictionary<Direction, Point> PositionalAdjustments = new Dictionary<Direction, Point>() 
        { 
            { Direction.North, new Point(0, 1) },
            { Direction.South, new Point(0, -1) },
            { Direction.East, new Point(1, 0) },
            { Direction.West, new Point(-1, 0) }
        };

        public Size Bounds { get; private set; }
        public Point CenterOfThePlanet { get; private set; }

        private readonly List<Obstacle> _obstacles;
        public IReadOnlyList<Obstacle> Obstacles
        {
            get { return _obstacles; }
        }

        public Rover()
            : this(new Size(25, 25))
        { }

        public Rover(Size bounds)
        {
            CenterOfThePlanet = new Point(Bounds.Width / 2, Bounds.Height / 2);
            LandOnMars(CenterOfThePlanet);
            Bounds = bounds;
            _obstacles = new List<Obstacle>();
        }

        private void LandOnMars(Point landingPoint)
        {
            if (!IsValidPosition(landingPoint))
                throw new CrashException("Doh! We tried to land on something other than the planet and the rover was destroyed!!!");
            Location = landingPoint;
            Facing = Direction.North;
        }

        public bool MoveBackward()
        {
            return Move(-1);
        }

        public bool MoveForward()
        {
            return Move(1);
        }

        private bool Move(int adjustmentFactor)
        {
            var desiredPosition = CreateDesiredPosition(adjustmentFactor);
            var newLocation = CalculateFinalPosition(Location, desiredPosition);

            var success = Location != newLocation;

            Location = newLocation;

            return success;
        }

        private Point CreateDesiredPosition(int adjustmentFactor)
        {
            var adjustment = PositionalAdjustments[Facing] * adjustmentFactor;
            return Location + adjustment;
        }

        public bool TurnRight()
        {
            if (Facing == Direction.North)
                Facing = Direction.East;
            else if (Facing == Direction.South)
                Facing = Direction.West;
            else if (Facing == Direction.East)
                Facing = Direction.South;
            else if (Facing == Direction.West)
                Facing = Direction.North;

            return true;
        }

        public bool TurnLeft()
        {
            if (Facing == Direction.North)
                Facing = Direction.West;
            else if (Facing == Direction.South)
                Facing = Direction.East;
            else if (Facing == Direction.East)
                Facing = Direction.North;
            else if (Facing == Direction.West)
                Facing = Direction.South;
            
            return true;
        }

        public void AddObstacle(int x, int y)
        {
            Point location = new Point(x, y);
            var obstacle = new Obstacle(location);
            _obstacles.Add(obstacle);
        }

        public Point CalculateFinalPosition(Point from, Point desired)
        {
            Point newDestination = desired;
            newDestination = CalculatePositionY(desired, newDestination);
            newDestination = CalculatePositionX(desired, newDestination);

            if (!IsValidPosition(newDestination))
            {
                return from;
            }

            return newDestination;
        }

        private Point CalculatePositionX(Point desired, Point newDestination)
        {
            if (desired.X > Bounds.Width)
            {
                newDestination = new Point(0, desired.Y);
            }
            if (desired.X < 0)
            {
                newDestination = new Point(Bounds.Width, desired.Y);
            }
            return newDestination;
        }

        private Point CalculatePositionY(Point desired, Point newDestination)
        {
            if (desired.Y > Bounds.Height)
            {
                newDestination = new Point(desired.X, 0);
            }
            if (desired.Y < 0)
            {
                newDestination = new Point(desired.X, Bounds.Height);
            }
            return newDestination;
        }

        public bool IsValidPosition(Point point)
        {
            bool anyInstance = _obstacles.Any(x => x.Location.Equals(point));

            return !anyInstance;
        }
    }
}

File: SRP/Java_Exercise/MarsRoverKata/src/main/java/MarsRoverKata/Rover.java
package MarsRoverKata;

import com.google.common.base.Predicate;
import com.google.common.collect.Collections2;
import com.google.common.collect.Iterables;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Rover {
    private Point location;
    private Direction facing;

    private Size bounds; // get
    private Point centerOfThePlanet; // get

    public Size getBounds() {
        return bounds;
    }

    private final List<Obstacle> obstacles;

    public List<Obstacle> getObstacles() {
        return obstacles;
    }

    public Rover() throws CrashException {
        this(new Size(25, 25));
    }

    public Rover(Size bounds) throws CrashException {
        this.bounds = bounds;
        centerOfThePlanet = new Point(this.bounds.getWidth() / 2, this.bounds.getHeight() / 2);
        obstacles = new ArrayList<Obstacle>();
        setFacing(Direction.North);
        landOnMars(centerOfThePlanet);
    }

    public Point getLocation() { return location; }

    public void setLocation(Point location) {
        this.location = location;
    }

    public Direction getFacing() { return facing; }

    public void setFacing(Direction facing) {
        this.facing = facing;
    }

    private static final Map<Direction, Point> positionalAdjustments = new HashMap<Direction, Point>() {{
        put(Direction.North, new Point(0, 1));
        put(Direction.South, new Point(0, -1));
        put(Direction.East, new Point(1, 0));
        put(Direction.West, new Point(-1, 0));
    }};


    private void landOnMars(Point landingPoint) throws CrashException {
        if (!isValidPosition(landingPoint)) {
            throw new CrashException("Doh! We tried to land on something other than the planet and the rover was destroyed!!!");
        }
        setLocation(landingPoint);
        setFacing(Direction.North);
    }

    public boolean moveBackward() {
        return move(-1);
    }

    public boolean moveForward()
    {
        return move(1);
    }

    private boolean move(int adjustmentFactor) {
        Point desiredPosition = createDesiredPosition(adjustmentFactor);
        Point newLocation = calculateFinalPosition(getLocation(), desiredPosition);
        boolean success = getLocation() != newLocation;
        setLocation(newLocation);
        return success;
    }

    private Point createDesiredPosition(int adjustmentFactor) {
        Point adjustment = Point.multiply(positionalAdjustments.get(getFacing()), adjustmentFactor);
        return Point.add(getLocation(), adjustment);
    }

    public boolean turnRight()
    {
        if (getFacing() == Direction.North)
            setFacing(Direction.East);
        else if (getFacing() == Direction.South)
            setFacing(Direction.West);
        else if (getFacing() == Direction.East)
            setFacing(Direction.South);
        else if (getFacing() == Direction.West)
            setFacing(Direction.North);

        return true;
    }

    public boolean turnLeft()
    {
        if (getFacing() == Direction.North)
            setFacing(Direction.West);
        else if (getFacing() == Direction.South)
            setFacing(Direction.East);
        else if (getFacing() == Direction.East)
            setFacing(Direction.North);
        else if (getFacing() == Direction.West)
            setFacing(Direction.South);

        return true;
    }

    public void addObstacle(Obstacle obstacle) {
        obstacles.add(obstacle);
    }

    public Point calculateFinalPosition(Point from, Point desired) {
        Point newDestination = desired;
        newDestination = calculatePositionY(desired, newDestination);
        newDestination = calculatePositionX(desired, newDestination);

        if (!isValidPosition(newDestination)) {
            return from;
        }

        return newDestination;
    }

    private Point calculatePositionX(Point desired, Point newDestination) {
        if (desired.getX() > bounds.getWidth()) {
            newDestination = new Point(0, desired.getY());
        }
        if (desired.getX() < 0) {
            newDestination = new Point(bounds.getWidth(), desired.getY());
        }
        return newDestination;
    }

    private Point calculatePositionY(Point desired, Point newDestination) {
        if (desired.getY() > bounds.getHeight()) {
            newDestination = new Point(desired.getX(), 0);
        }
        if (desired.getY() < 0) {
            newDestination = new Point(desired.getX(), bounds.getHeight());
        }
        return newDestination;
    }

    public boolean isValidPosition(final Point point) {
        boolean anyInstance = Iterables.any(obstacles, new Predicate<Obstacle>() {
            @Override
            public boolean apply(Obstacle input) {
                return input.getLocation().equals(point);
            }
        });
        return !anyInstance;
    }
}

